pkg/provisioner/crdprovisioner.go

```diff
@@ -375,72 +380,60 @@ func (c *CrdProvisioner) PublishVolume(
+	if azVolume.Spec.MaxMountReplicaCount > 0 && len(existingAttachments) > azVolume.Spec.MaxMountReplicaCount {
+		roleChangeLabel, _ := azureutils.CreateLabelRequirements(consts.RoleChangeLabel, selection.Equals, consts.Demoted)
+		demotedAttachments, err := azVALister.List(azVASelector.Add(*roleChangeLabel))
+		// if there is no demoted AzVolumeAttachment, randomly pick a replica to unpublish (TODO: come up with a logic to select replica for unpublishing)
+		var attachmentToUnpublish *diskv1alpha2.AzVolumeAttachment
+		if err != nil || len(demotedAttachments) == 0 {
+			attachmentToUnpublish = existingAttachments[0]
+		} else {
+			attachmentToUnpublish = demotedAttachments[0]
 		}
-		// If there already exists a primary attachment with populated responseObject return
-		if azVolumeAttachmentInstance.Status.Detail != nil && azVolumeAttachmentInstance.Status.Detail.PublishContext != nil && azVolumeAttachmentInstance.Status.Detail.Role == diskv1alpha2.PrimaryRole {
-			return azVolumeAttachmentInstance.Status.Detail.PublishContext, nil
-			// if attachment is pending return error to prevent redundant attachment request
-		} else if azVolumeAttachmentInstance.Status.State == diskv1alpha2.Attaching {
-			return nil, status.Errorf(codes.Aborted, "attachment still in process for volume (%s) and node (%s)", volumeName, nodeID)
+
+		if err := c.UnpublishVolume(ctx, attachmentToUnpublish.Spec.VolumeID, attachmentToUnpublish.Spec.NodeName, nil); err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to make request to unpublish volume (%s) from node (%s): %v", attachmentToUnpublish.Spec.VolumeName, attachmentToUnpublish.Spec.NodeName, err)
 		}
+	}
 
-		updateFunc := func(obj interface{}) error {
-			updateInstance := obj.(*diskv1alpha2.AzVolumeAttachment)
-			if updateInstance.Status.Error != nil {
-				// reset error and state field of the AzVolumeAttachment so that the reconciler can retry attachment
-				updateInstance.Status.Error = nil
-				updateInstance.Status.State = diskv1alpha2.AttachmentPending
-			}
+	attachmentName := azureutils.GetAzVolumeAttachmentName(volumeName, nodeID)
 
-			// Otherwise, we are trying to update
-			// the updateInstanceolumeAttachment role from Replica to Primary
-			updateInstance.Spec.RequestedRole = diskv1alpha2.PrimaryRole
-			// Keeping the spec fields up to date with the request parameters
-			updateInstance.Spec.VolumeContext = volumeContext
-			// Update the label of the AzVolumeAttachment
-			if updateInstance.Labels == nil {
-				updateInstance.Labels = map[string]string{}
-			}
-			updateInstance.Labels[consts.RoleLabel] = string(diskv1alpha2.PrimaryRole)
+	attachmentObj, err := azVALister.Get(attachmentName)
 
-			return nil
+	// if replica AzVolumeAttachment CRI does not exist for the volume-node pair, create one
+	if errors.IsNotFound(err) {
+		// check if number of volumeAttachments for the volume has met the maxShares limit or not.
+		if err != nil && !errors.IsNotFound(err) {
```

Comment:
The check on line 431 guarantees that `err != nil` and `errors.IsNotFound(err) == true`, so we would never execute the body of the `if` statement. What was your intent here?

Reply:
Good catch. I think it got misplaced in process of rebasing and resolving merge conflicts.